src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,54 @@
+#include "firmware/shared/circular_buffer.h"
+
+void circular_buffer_init(struct circular_buffer *cbuff)
+{
+    cbuff->head   = 0;
+    cbuff->tail   = 0;
+    cbuff->isFull = false;
+    // Initialize buffer size to CIRCULAR_BUFFER_MAX_NUM_ELEMENTS
+    cbuff->bufferData = (float *)malloc(sizeof(float) * CIRCULAR_BUFFER_MAX_NUM_ELEMENTS);
+}
+
+void circular_buffer_free(struct circular_buffer *cbuff)
+{
+    free(cbuff->bufferData);
+}
+
+// Note: Implementation assumes head and tail are equal when full
+void circular_buffer_push(struct circular_buffer *cbuff, float data)
+{
+    cbuff->bufferData[cbuff->head % CIRCULAR_BUFFER_MAX_NUM_ELEMENTS] = data;
+    cbuff->head++;
+    if (cbuff->head > CIRCULAR_BUFFER_MAX_NUM_ELEMENTS)
+    {
+        cbuff->tail++;
+    }
+    if (cbuff->head == cbuff->tail)
+    {
+        cbuff->isFull = true;
+    }
+}
+
+// Note: Index is starting from 1 for most recent
+float circular_buffer_get_index(struct circular_buffer *cbuff, int index)
+{
+    // Check if index is larger than the maximum buffer size
+    if (index > CIRCULAR_BUFFER_MAX_NUM_ELEMENTS)
```

Comment:
lets `assert` here so we hard fault instead of retuning NAN, might be safer because NAN has weird behavior 

see: https://stackoverflow.com/questions/1923837/how-to-use-nan-and-inf-in-c

essentially if you do `float a = NAN;` and then try to do `a == NAN` it evaluates to false, we wouldn't be able to properly recover here

Reply:
Added `assert` statements to both.